Palette switching

knguye23 opened the pull request:
# Pull Request Description

Please fill out the form below. For the checkbox, use [x] to have a checkmark. For example:

- [x] With a checkmark.
- [ ] Without a checkmark.

This should not affect the way whether your pull request is accepted or not. Think of it like a survey and documentation, so the reviewers and future developers can follow your PR easier.

1. Have you included *.gitignore* file? Or manually remove config and environment files (E.g: *node_modules* folder, Eclipse IDE files) from this submission? The answer should be **Yes** in this case.

   - [x] Yes
   - [ ] No

2. Did you fully complete the assigned task(s)?

   - [x] Fully completed
   - [ ] Partially completed

3. Did you introduce a new non-required feature in this PR? This feature is not an assigned task, but it is the one that you feel will make the game better. For example, reproduce a bug used by speed runner in an original version.

   - [ ] Yes
   - [x] No

4. Are you aware of any bug caused by your code that have not been fixed in this PR?

   - [ ] Yes
   - [x] No

5. Are you aware of any **potential bug**? You don't have to fix this, it is just a way to document. You may not face the bug during the development process, but theoretically it can happens in edge cases. For example, memory leak or a player falls out of map.

   - [ ] Yes
   - [x] No

## Unimplented task(s)

- N/A

## Related issue(s)

- 17

## Added

- Swapped palette on "stuff.png" and "bricksolod.png" to match underground palette.

## Changed and Enhanced

- N/A

## Fixed

- N/A

## Bugs

1. Not-fixed bugs (only answer when you chose **Yes** in question 4):
   - N/A
2. Potential bugs (only answer when you chose **Yes** in question 5):
   - N/A

## Contributors and code author(s)

- Colin Chan (colinzzz43) and Khue Nguyen (knguye23)

## Final check

Have you fully filled out this document, and add a label **PR: NEED REVIEW** to this pull request? Answer *Yes* (as it should be) means you are good to go! Thanks for the contribution.

- [x] Yes
- [ ] No

hunghvu commented:
@algorithm0r, @knguye23, @colinzzz43 For this PR, Khue contacted me before and let me know the PR will only introduce new sprites without any code, even though an issue requires new code/feature. I let her decide on this matter/approach herself. Because the nature of this PR is different, I think it's necessary to let Chris know and review this PR. 

By the way, this PR does not ignore VS Code file. This need to be changed.

BenjaminDeJager reviewed (approved):
no code, potentially a non-pr.
hung says he wants to let chris decide so why not. (I am approving it, but that doesn't mean merging after all)

algorithm0r reviewed (dismissed):
This should include code changes to the brick class to allow for display of the palette switched entities.  This is probably easily accomplished with a single Boolean variable added to the constructor to determine if the underground bricks should be used or not.  I do this for Luigi so you can check Mario.js for an example.

Update (3/12 by @hunghvu): This PR should satisfy the request now, so the request will be dismissed.

BenjaminDeJager reviewed (approved):
not actually tested, but I see no issues with the code itself atm.
underground_bricksolod seems to be misspelled "solid" however.

BenjaminDeJager reviewed (approved):
not actually tested, but I see no issues with the code itself atm.
underground_bricksolod seems to be misspelled "solid" however.

edit: possably "approved" the wrong one, will see....

BenjaminDeJager commented:
FYI: I resolved a conflict in brick.js by pushing to colin's branch,

hunghvu reviewed (changes requested):
Game is not runnable due to reference error. Still, I think it's necessary to have a decision on this PR due to the nature of it.
![image](https://user-images.githubusercontent.com/62678592/109543765-8ff9a200-7a7b-11eb-994c-ecc2e48baa48.png)

hunghvu commented on bricks.js:
```diff
@@ -1,9 +1,12 @@
 
         this.spritesheet = ASSET_MANAGER.getAsset("./sprites/bricks.png");
 
+        if (underground) this.spritesheet = ASSET_MANAGER.getAsset("./sprites/SuperMarioBros/sprites/underground_stuff.png");
```
This and brick, invalid directory and will crash game upon being drawn (TypeError).

hunghvu reviewed (changes requested):
Change is requested. Please look at the comment above.

colinzzz43 reviewed (commented):
sorry just got it fixed. I have cloned the forked repo and this repo to my local machine, so when I was merging to the master, I didn't know it wasn't merging to the master branch in the forked repo.

hunghvu commented:
@colinzzz43 May I ask the reason you remove *underground* flag. From what I see, the flag itself is not a problem, but an asset link/directory is. The current sprite sheet directory you are providing is invalid so an image is undefined and cannot be drawn. 

Besides, in JavaScript, parameters are optional so you can pass as many arguments as you want. In this case, you introduce a new flag to constructor, which is *underground*. This means you change the prototype of that object, which is not recommended. For example, you have a new flag *underground* in constructors, certainly, the caller on master branch won't have this flag. In JS, that's fine as the reason mentioned above. Unprovided *underground* is default to be undefine. Personally, I will avoid *undefine* as much as possible and provide a default parameter there (false). If this happens in a compiled language, there is a high chance you will face compiled error (depends on language rules, e.g: Java will result in compiled error).

hunghvu reviewed (changes requested):
Change is requested, the mentioned issue has not been fixed. In your next update, please also let us know the way to demo your code.

colinzzz43 reviewed (commented):
First time using this feature in Github, the changes didn't go through for some reason. Will see if it goes through this time, sorry if I'm bothering anyone.

hunghvu commented:
@colinzzz43 Can you let me know the way to demo your PR? I tweak it myself, but it's better to follow your instruction to reduce unnecessary noise. I will look at this PR tomorrow.

Update as of 9:30 am Mar. 2, 2021: There is no instruction on how to modify the code for testing purpose so I cannot continue. Meanwhile, @BenjaminDeJager, can you look at the code and recommend some changes if necessary? In my experience, this PR is quite buggy.

knguye23 commented:
@hunghvu Hi Hung, I added a boolean property called _underground_ in each of the level objects under _levels.js_. Since level 1-1 does not use the palette swapped spritesheet, I set _underground_ to be false.
https://github.com/colinzzz43/SuperMarioBros/blob/0f757403a88ad04a2c519783932ead3e86365b62/levels.js#L1-L5
https://github.com/colinzzz43/SuperMarioBros/blob/0f757403a88ad04a2c519783932ead3e86365b62/levels.js#L109-L112

However, level 1-2 is set in the underworld so _underground_ should be true. I only included the following comment block for demonstration purposes since I didn't have the code for level 1-2's map layout. When you're testing level 1-2, just add `underground: true` as demonstrated below.
https://github.com/colinzzz43/SuperMarioBros/blob/0f757403a88ad04a2c519783932ead3e86365b62/levels.js#L181-L189

I also added a way to change the canvas background color to black when a level is set in the underworld. However, there is a still a bug. When Mario dies, the game transitions to level 1-1 (which is not an underworld level) but the canvas background color stays black instead of changing to the sky blue color.
https://github.com/colinzzz43/SuperMarioBros/blob/0f757403a88ad04a2c519783932ead3e86365b62/sceneManager.js#L183-L189

algorithm0r commented:
We could test the underground bricks in the level one bonus level.  I have that already implemented in levels.js.

hunghvu reviewed (dismissed):
Well done, it all works now. I will approve this PR. Regarding the bug that Khue mentioned, I made the following video to have a small discussion on it. The video is cut down to meet 10MB limit of GitHub, but that's enough anyway.


https://user-images.githubusercontent.com/62678592/109750236-e99fc080-7b90-11eb-9286-8422a8fe548c.mp4

BenjaminDeJager commented on levels.js:
```diff
@@ -173,4 +175,15 @@ var bonusLevelOne = {
\ No newline at end of file
+};
+
+// NOTE: THIS IS JUST A CODE SNIPPET.
```
good to leave a meta-comment explaining this.

BenjaminDeJager reviewed (approved):
as far as I see, this covers the issues that were previously brought up by hung and he's approved so I'm OK-ing this.

BenjaminDeJager reviewed (dismissed):
it seems that ground, brick, etc have had the requested boolean added.

BenjaminDeJager commented:
it seems I accidentally dismissed hungs approval by doing that conflict resolution @hunghvu please re-add your approval, maybe double check that I resolved the conflict correctly cas its not letting me push the conflict-fix in any direction so I can't test it.

hunghvu commented on levels.js:
```diff
@@ -177,13 +177,287 @@ var bonusLevelOne = {
     var levelTwo = {
     underground: true, // NOTE: PLEASE ADD THIS BOOLEAN PROPERTY WHEN TESTING LEVEL 1-2
+
+var levelTwo = {
```
Due to bad merge, this redundant line 183 will break the game. @BenjaminDeJager Please fix the issue, then we can merge.

BenjaminDeJager reviewed (dismissed):
trying to fix merge-mistake

BenjaminDeJager reviewed (approved):
should be good to go now, removed redundent line in levels.js

hunghvu commented:
The merging issue is now fixed and it's ready to be merged now.

BenjaminDeJager commented:
woot woot! 

done!
